python/jupyter/src/widget.ts

```diff
@@ -92,6 +112,100 @@ class ChemiscopeBaseView extends DOMWidgetView {
             this.warnings.defaultTimeout = timeout;
         }
     }
+
+    /**
+     * Install a handler for custom messages coming from Python
+     */
+    protected _enableStructureMessages(): void {
+        this.model.on(
+            'msg:custom',
+            (content: StructureRequest, _buffers: unknown[]) => {
+                void _buffers; // explicitly mark as intentionally unused
+                if (!content || typeof content !== 'object') {
+                    return;
+                }
+
+                if (content.type === 'load-structure-result') {
+                    const requestId = content.requestId;
+                    const pending = this._pendingStructureRequests.get(requestId);
+                    if (pending) {
+                        this._pendingStructureRequests.delete(requestId);
+                        // python can send either a string (JSON-serialized) or an object
+                        const raw = content.structure;
+                        const structure: Structure =
+                            typeof raw === 'string'
+                                ? (JSON.parse(raw) as Structure)
+                                : (raw as Structure);
+                        pending.resolve(structure);
+                    }
+                } else if (content.type === 'load-structure-error') {
+                    const requestId = content.requestId;
+                    const pending = this._pendingStructureRequests.get(requestId);
+                    if (pending) {
+                        this._pendingStructureRequests.delete(requestId);
+                        pending.reject(
+                            new Error(content.error || 'unknown error while loading structure')
+                        );
+                    }
+                    if (content.error) {
+                        this.warnings.sendMessage(`Error loading structure: ${content.error}`);
+                    }
+                }
+            },
+            this
+        );
+    }
+
+    /**
+     * Ask Python to load one structure given the user-defined `data`.
+     * `data` is typically the filename or an object containing it.
+     */
+    protected _requestStructureFromPython(index: number, filename: string): Promise<Structure> {
+        const requestId = this._nextRequestId++;
+        return new Promise<Structure>((resolve, reject) => {
+            this._pendingStructureRequests.set(requestId, { resolve, reject });
+
+            this.model.send({
+                type: 'load-structure',
+                requestId,
+                index,
+                filename,
```

Comment:
I think this should just send the `data` from UserStructure as-is (instead of sending "filename"), so we can change it later more easily

Reply:
I renamed it in a more generic way, but we can't send any type, so I think this is it. Feel free to change it if you want something else, I don't understand what you want here